system/include/SSE/xmmintrin.h

```diff
@@ -0,0 +1,709 @@
+_mm_set_ps(float __z, float __y, float __x, float __w)
+{
+  return wasm_f32x4_make(__w, __x, __y, __z);
+}
+
+static __inline__ __m128 __attribute__((__always_inline__))
+_mm_setr_ps(float __z, float __y, float __x, float __w)
+{
+  return wasm_f32x4_make(__z, __y, __x, __w);
+}
+
+static __inline__ __m128 __attribute__((__always_inline__))
+_mm_set_ss(float __w)
+{
+  return wasm_f32x4_make(__w, 0, 0, 0);
+}
+
+static __inline__ __m128 __attribute__((__always_inline__))
+_mm_set_ps1(float __w)
+{
+  return wasm_f32x4_splat(__w);
+}
+
+#define _mm_set1_ps _mm_set_ps1
+
+static __inline__ __m128 __attribute__((__always_inline__))
+_mm_setzero_ps(void)
+{
+  return wasm_f32x4_const(0.f, 0.f, 0.f, 0.f);
+}
+
+static __inline__ __m128 __attribute__((__always_inline__))
+_mm_load_ps(const float *__p)
+{
+  return *(__m128*)__p;
+}
+
+static __inline__ __m128 __attribute__((__always_inline__))
+_mm_loadl_pi(__m128 __a, const void /*__m64*/ *__p)
+{
+  return __builtin_shufflevector(wasm_f32x4_make(((float*)__p)[0], ((float*)__p)[1], 0.f, 0.f), __a, 0, 1, 6, 7);
+}
+
+static __inline__ __m128 __attribute__((__always_inline__))
+_mm_loadh_pi(__m128 __a, const void /*__m64*/ *__p)
+{
+  return __builtin_shufflevector(__a, wasm_f32x4_make(((float*)__p)[0], ((float*)__p)[1], 0.f, 0.f), 0, 1, 4, 5);
+}
+
+static __inline__ __m128 __attribute__((__always_inline__))
+_mm_loadr_ps(const float *__p)
+{
+  __m128 __v = _mm_load_ps(__p);
+  return __builtin_shufflevector(__v, __v, 3, 2, 1, 0);
+}
+
+static __inline__ __m128 __attribute__((__always_inline__))
+_mm_loadu_ps(const float *__p)
+{
+  return wasm_v128_load(__p);
+}
+
+static __inline__ __m128 __attribute__((__always_inline__))
+_mm_load_ps1(const float *__p)
+{
+  return wasm_v32x4_load_splat(__p);
+}
+#define _mm_load1_ps _mm_load_ps1
+
+static __inline__ __m128 __attribute__((__always_inline__))
+_mm_load_ss(const float *__p)
+{
+  return wasm_f32x4_make(*__p, 0.f, 0.f, 0.f);
+}
+
+static __inline__ void __attribute__((__always_inline__))
+_mm_storel_pi(__m64 *__p, __m128 __a)
+{
+  *__p = (__m64)__a;
+}
+
+static __inline__ void __attribute__((__always_inline__))
+_mm_storeh_pi(__m64 *__p, __m128 __a)
+{
+  *__p = (__m64)__builtin_shufflevector(__a, __a, 6, 7, 2, 3); // _mm_movehl_ps pattern
+}
+
+static __inline__ void __attribute__((__always_inline__))
+_mm_store_ps(float *__p, __m128 __a)
+{
+  *(__m128 *)__p = __a;
+}
+// No NTA cache hint available.
+#define _mm_stream_ps _mm_store_ps
+
+#define _MM_HINT_T0 3
+#define _MM_HINT_T1 2
+#define _MM_HINT_T2 1
+#define _MM_HINT_NTA 0
+// No prefetch available, dummy it out.
+static __inline__ void __attribute__((__always_inline__))
+_mm_prefetch(const void *__p, int __i)
+{
+  ((void)__p);
+  ((void)__i);
+}
+
+static __inline__ void __attribute__((__always_inline__))
+_mm_sfence(void)
+{
+  __sync_synchronize(); // Emscripten/SharedArrayBuffer has only a full barrier instruction, which gives a stronger guarantee.
```

Comment:
It would make more sense for this comment to say that WebAssembly's memory model is sequentially consistent for now, and that this implementation could be weakened in the future once the memory model is extended to include store-release semantics.

Reply:
Check